docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -294,254 +296,285 @@ TODO
+handshake.
+
+### Internal data structures
+
+In the mempool, a lane is defined by its priority:
+```golang
+type Lane uint32
+```
+
+Currently, the `CListMempool` data structure has two fields to store and access transactions:
+```golang
+txs    *clist.CList // Concurrent list of mempool entries.
+txsMap sync.Map     // Map of type TxKey -> *clist.CElement, for quick access to elements in txs.
+```
+
+With the introduction of lanes, the main change will be to divide the `CList` data structure into
+$N$ `Clist`s, one per lane. `CListMempool` will have the following fields:
+```golang
+lanes   map[Lane]*clist.CList
+txsMap  sync.Map // Map of type TxKey -> *clist.CElement, for quick access to elements in lanes.
+txLanes sync.Map // Map of type TxKey -> Lane, for quick access to the lane corresponding to a tx.
+
+// Fixed variables set during initialization.
+defaultLane Lane
+sortedLanes []Lane // Lanes sorted by priority
+```
+The auxiliary fields `txsMap` and `txLanes` are, respectively, for direct access to the mempool
+entries, and for direct access to the lane of a given transaction.
+
+If the application does not implement lanes (that is, it responds with empty values in
+`InfoResponse`), then `defaultLane` will be set to 1, and `lanes` will have only one entry for the
+default lane.
+
+`CListMempool` also contains the cache, which is only needed before transactions have a lane
+assigned. Since the cache is independent of the lanes, we do not need to modify it.
+
+### Configuration
+
+For an MVP, all lanes can share the same mempool configuration. In this scenario, all lanes will be
+capped by the total mempool capacities as currently defined in the configuration. Namely, these are:
+- `Size`, the total number of transactions allowed in the mempool, 
+- `MaxTxsBytes`, the maximum total number of bytes of the mempool, and
+- `MaxTxBytes`, the maximum size in bytes of a single transaction accepted into the mempool.
+
+With lanes, the total size of the mempool will be the sum of the sizes of all lanes.
+
+Additionally, the `Recheck` and `Broadcast` flags will apply to all lanes or to none. Remember that,
+with `PrepareProposal`, it becomes _always_ mandatory to recheck remaining transactions in the
+mempool, so there is no point in disabling `Recheck` per lane.
+
+### Adding transactions to the mempool
+
+When validating a transaction received for the first time with `CheckTx`, the application will
+optionally return its lane in the response.
+```protobuf
+message CheckTxResponse {
+  ...
+  uint32 lane = 12;
+}
+```
+The callback that handles the first-time CheckTx response will append the new mempool entry to the
+corresponding `CList`, namely `lanes[lane]`, and update the other auxiliary variables accordingly.
+If `lane` is 0, it means that the application did not set any lane in the response message, so the
+transaction will be assigned to the default lane.
+
+### Removing transactions from the mempool
```

Comment:
Should this also be present in the list at the beginning of the section?

Reply:
I didn't add it because there's nothing important about removing transactions.